SparseGrids/gridtestExternalTests.cpp

```diff
@@ -215,101 +262,125 @@ TestResults ExternalTester::getError(const BaseFunction *f, TasGrid::TasmanianSp
-        for(int i=0; i<num_mc; i++){
-            grid.evaluate(&(test_x[i * num_dimensions]), &(result_tasm[i * num_outputs]));
-            f->eval(&(test_x[i * num_dimensions]), &(result_true[i * num_outputs]));
-        }
+            for(int i=0; i<num_mc; i++){
+                for(int k=0; k<num_entries; k++){
+                    double nrmik = std::abs(result_true[i * num_entries + k]);
+                    double errik = std::abs(result_true[i * num_entries + k] - result_tasm[i * num_entries + k]);
+                    if (nrm[k] < nrmik) nrm[k] = nrmik;
+                    if (err[k] < errik) err[k] = errik;
+                }
+            }
 
-        for(int i=0; i<num_mc; i++){
-            for(int k=0; k<num_outputs; k++){
-                double nrmik = std::abs(result_true[i * num_outputs + k]);
-                double errik = std::abs(result_true[i * num_outputs + k] - result_tasm[i * num_outputs + k]);
-                if (nrm[k] < nrmik) nrm[k] = nrmik;
-                if (err[k] < errik) err[k] = errik;
+            double rel_err = 0.0; // relative error
+            for(int k=0; k<num_entries; k++){
+                double relative_errork = std::fabs(nrm[k]) <= Maths::num_tol ? err[k] : err[k] / nrm[k];
+                if (rel_err < relative_errork) rel_err = relative_errork;
             }
-        }
 
-        double rel_err = 0.0; // relative error
-        for(int k=0; k<num_outputs; k++){
-            double relative_errork = err[k] / nrm[k];
-            if (rel_err < relative_errork) rel_err = relative_errork;
-        }
+            if (type == type_internal_differentiation)
+                rel_err = std::max(rel_err, unitDerivativeTests(f, grid));
 
-        R.error = rel_err;
+            R.error = rel_err;
+        }
     }
     R.num_points = grid.getNumPoints();
     return R;
 }
 
-bool ExternalTester::testGlobalRule(const BaseFunction *f, TasGrid::TypeOneDRule rule, const int *anisotropic, double alpha, double beta, const bool interpolation, const int depths[], const double tols[]) const{
+bool ExternalTester::testGlobalRule(const BaseFunction *f, TasGrid::TypeOneDRule rule, const int *anisotropic, double alpha, double beta, const std::vector<TestType> &tests, const int depths[], const double tols[]) const{
     TasGrid::TasmanianSparseGrid grid;
     TestResults R;
-    int num_global_tests = (interpolation) ? 3 : 1;
-    TestType tests[3] = { type_integration, type_nodal_interpolation, type_internal_interpolation };
+    size_t num_global_tests = tests.size();
     TasGrid::TypeDepth type = (rule == rule_fourier ? TasGrid::type_level : TasGrid::type_iptotal);
     std::vector<double> x = genRandom(f->getNumInputs());
     if (rule == rule_fourier){ for(int i=0; i<f->getNumInputs(); i++) x[i] = 0.5*(x[i]+1.0); }    // map to canonical [0,1]^d
     bool bPass = true;
+    bool integration(false), interpolation(false), differentiation(false);
```

Comment:
You have these granular conditions of different possibilities, but do you need this?

Every time you're checking these variables you are checking for `(interpolation or differentiation)` together, and even in that case you're only using the test to determine the number of outputs.

Why not have a new variable called `num_outputs`, set it to zero and then loop over `tests` if you see a test with internal interpolation/differentiation, then you set `num_outputs = f->getNumOutputs()` and simplify the code below.

Reply:
Changed. I made a new variable called `num_fn_outputs` (to differentiate it from the grid's `num_outputs`) to store this information.